### PR TITLE
Feat: create extended ledger

### PR DESCRIPTION
**Pull request by AlexTheWizardL**
### This pull request covers changes from #60 issue to create separate for the "hackish" code.

The PR is best reviewed commit by commit.

- d3e68af7ec5cf7415cd61a083b4ac5ffa199fe40 create `CashCtrlLedgerExtended` class
- 5cfdca38f74b0896a85b52b479b2c0615d7a6231 move "hackish" code and update test suits to use new class
- 5316d88f9214b9b1fc99ae6295a651bd67475660 update docstrings for `CashCtrlLedger` and `CashCtrlLedgerExtended` classes
- d04b3ee5602959aa532662cc354fb37f115c1632 update readme with **Core Components** and **Key Features** sections

@lasuk Please review the changes and provide feedback or approval for merging.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,10 +2,39 @@
...
+`cashctrl_ledger` is a Python package that implements the `pyledger.LedgerEngine`
+interface, enabling seamless integration with the CashCtrl accounting service. With this package,
+users can perform various accounting operations programmatically, directly from Python.
+
+## Core Components
+
+1. **CashCtrlLedger**\
+The class that Implements the `pyledger` interface by connecting to the CashCtrl online accounting software via REST API.
+
+2. **CashCtrlLedgerExtended**\
```
Comment (lasuk) [inline README.md:14]: I am not convinced by name and description of the new class. In one sentence, why do we need this class? What does it do? What is so cool about it?

File: README.md [hunk 2]
```diff
@@ -2,10 +2,39 @@
...
+
+- **Ledger Management:**
+  - Retrieve and standardize ledger entries from the CashCtrl account.
+  - Add, update, and delete ledger entries in the CashCtrl account.
+  - Mirror ledger data between local and remote states.
+  - Manage ledger entry attachments based on specified file paths.
+
+- **Currency Management:**
+  - Retrieve the base currency of the CashCtrl account.
+  - Get precision for specific currencies (currently supports USD, CHF, EUR).
```
Comment (lasuk) [inline README.md:37]: The description of key features is redundant and not particularly helpful. Can you think of three to five sentences to describe the features to someone who is not familiar with Pyledger nor with CashCtrl?

File: cashctrl_ledger/__init__.py [hunk 3]
```diff
@@ -5,10 +5,13 @@
 package, users can perform various accounting operations programmatically, directly from Python.
 
 Modules:
-- ledger: Contains the CashCtrlLedger class to sync ledger system onto CashCtrl.
+- ledger: Contains the CashCtrlLedger Class that Implements the pyledger interface by connecting
+          to the CashCtrl online accounting software.
+- extended_ledger: Contains the CashCtrlLedgerExtended class that violates limitations of the CashCtrl.
```
Comment (lasuk) [inline cashctrl_ledger/__init__.py:10]: "that violates limitations of the CashCtrl" is not correct.

File: cashctrl_ledger/extended_ledger.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
+"""This module contains "hackish" code that violates the limitations of the CashCtrl."""
```
Comment (lasuk) [inline cashctrl_ledger/extended_ledger.py:1]: Poor docstring.  In one sentence, why do we need this class? What does it do?

File: cashctrl_ledger/extended_ledger.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+from typing import Union
+import numpy as np
+import pandas as pd
+from .ledger import CashCtrlLedger
+
+
+class CashCtrlLedgerExtended(CashCtrlLedger):
+    """Class that inherits the CashCtrlLedger class and violates
+    the limitations of the CashCtrl.
```
Comment (lasuk) [inline cashctrl_ledger/extended_ledger.py:11]: "violates the limitations of the CashCtrl." is not correct.

File: cashctrl_ledger/extended_ledger.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+    @transitory_account.setter
+    def transitory_account(self, value: int):
+        self._transitory_account = value
+
+    # ----------------------------------------------------------------------
+    # Ledger
+
+    def sanitize_ledger(self, ledger: pd.DataFrame) -> pd.DataFrame:
+        """Sanitizes the ledger DataFrame by splitting multi-currency transactions
+        and ensuring FX adjustments.
```
Comment (lasuk) [inline cashctrl_ledger/extended_ledger.py:70]: Description is not clear. What exactly does the method do? Maybe worthwhile to mention that we are overriding a generic method to comply with specific limitations of CashCtrl.

File: cashctrl_ledger/ledger.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-"""Module to sync ledger system onto CashCtrl."""
+"""Module that implements the pyledger interface by connecting to the CashCtrl"""
```
Comment (lasuk) [inline cashctrl_ledger/ledger.py:1]: 'the CashCtrl' is poor English. Also, we should not assume that 'the CashCtrl' is a known concept.

Suggestion:
```
"""Module that implements the pyledger interface by connecting to the CashCtrl API."""
```

**Review by lasuk — COMMENTED** [review]
Thanks for separating CashCtrl accessor / mutator functions from the 'hawkish' code that splits transactions that can not be represented in CashCtrl into multiple representable transactions.

I left a series of comments where I was not convinced of descriptions and documentations.

File: README.md (hunk 1 above)
Comment (AlexTheWizardL) [reply README.md:14]: What do you think about the `CashCtrlLedgerHackish` as a better name?

Comment (lasuk) [reply README.md:14]: CashCtrlLedgerHackish is a terrible name: just states  that the class is in some way improper, without stating what it actually does, what features it contains, or the reasoning why we created it in the first place.

Comment (AlexTheWizardL) [reply README.md:14]: #### Class Name Proposals generated with help of ChatGPT

  - [ ] **CashCtrlLedgerSplitter** - Emphasizes the splitting of transactions within the ledger.
  - [ ] **CashCtrlLedgerNormalizer** - Focuses on ensuring transactions conform to CashCtrl's standards.
  - [ ] **CashCtrlLedgerBalancer** - Highlights the balancing of ledger entries after splitting or adjustments.
  - [ ] **CashCtrlTransactionSplitter** - Indicates that the class specifically handles the splitting of transactions.
  - [ ] **CashCtrlTransactionBalancer** - Suggests a focus on balancing transactions, possibly after splitting.
  - [ ] **CashCtrlNormalizeSplitter** - Combines normalization and splitting, indicating both actions are central to the class's functionality.

  **All of these options are valid and align well with the class's intended purpose.**

  You may want to choose based on which aspect of the class’s functionality you want to emphasize most:

  - **Splitting:** `CashCtrlLedgerSplitter`, `CashCtrlTransactionSplitter`
  - **Normalizing:** `CashCtrlLedgerNormalizer`, `CashCtrlNormalizeSplitter`
  - **Balancing:** `CashCtrlLedgerBalancer`, `CashCtrlTransactionBalancer`

**Comment by lasuk** [comment]
I've pushed revised documentation and  docstrings as e88b8e7.
@AlexTheWizardL: Please souls if you disagree.

**Comment by lasuk** [comment]
From my point of view, the sole remaining question is the class name.

File: README.md (hunk 1 above)
Comment (lasuk) [reply README.md]: Also:
- CashCtrlCompliantLedger: This name emphasizes that the ledger is fully compliant with CashCtrl’s rules and restrictions. It suggests that the class ensures all transactions adhere to CashCtrl’s standards.
- EnhancedCashCtrlLedger: This name suggests that the class not only ensures compliance but also extends or improves the base functionality of the CashCtrlLedger. It indicates that this version offers something more than the standard ledger.

**Pull request merged**